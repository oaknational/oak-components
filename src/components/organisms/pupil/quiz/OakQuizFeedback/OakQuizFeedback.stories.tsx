import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakQuizFeedback } from "./OakQuizFeedback";

import { OakFlex } from "@/components/atoms";

const meta: Meta<typeof OakQuizFeedback> = {
  component: OakQuizFeedback,
  tags: ["autodocs"],

  argTypes: {
    answerFeedback: {
      control: {
        type: "text",
      },
    },
  },
  args: {
    feedback: "correct",
    answerFeedback: "Well done!",
  },
  decorators: [
    (Story) => (
      <OakFlex $flexDirection="column" $gap="space-between-m">
        <Story />
      </OakFlex>
    ),
  ],
};
export default meta;

type Story = StoryObj<typeof OakQuizFeedback>;

export const Default: Story = {
  render: (args) => <OakQuizFeedback {...args} />,
};

export const AllFeedback: Story = {
  render: (args) => (
    <>
      <OakQuizFeedback
        {...args}
        feedback="correct"
        answerFeedback="Well done!"
      />
      <OakQuizFeedback
        {...args}
        feedback="incorrect"
        answerFeedback="Correct answer: Flying Eye Books, Flying Eye, flying eye books, Flying eye books, FLYING EYE BOOKS"
      />
      <OakQuizFeedback
        {...args}
        feedback="partially-correct"
        answerFeedback="Correct answers: front cover, text title, blurb"
      />
    </>
  ),
  args: {
    answerFeedback: "Good work!",
  },
};

export const WithNoAnswerFeedback: Story = {
  args: {
    feedback: "partially-correct",
    answerFeedback: undefined,
  },
};
