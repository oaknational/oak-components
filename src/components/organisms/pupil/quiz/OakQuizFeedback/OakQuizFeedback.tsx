import React, { ReactNode } from "react";

import { OakBox, OakFlex, OakSpan } from "@/components/atoms";
import { OakRoundIcon } from "@/components/molecules";

export type OakQuizFeedbackProps = {
  /**
   * Feedback for an answer
   * `partially-correct` can be used to feedback on a multi-answer MCQ where some
   * correct options were not selected and/or some incorrect answers were selected
   */
  feedback: "correct" | "incorrect" | "partially-correct";
  /**
   * Some additional content to present with the feedback.
   * This is likely the correct answer(s) to the question or some praise.
   *
   * Only displayed when `feedback` has been applied.
   */
  answerFeedback?: ReactNode;
};

/**
 * Gives feedback after a question has been answered
 */
export const OakQuizFeedback = ({
  feedback,
  answerFeedback,
}: OakQuizFeedbackProps) => {
  let feedbackLabel: string;

  switch (feedback) {
    case "correct":
      feedbackLabel = "Correct";
      break;
    case "incorrect":
      feedbackLabel = "Incorrect";
      break;
    case "partially-correct":
      feedbackLabel = "Almost correct";
      break;
  }

  return (
    <OakBox aria-live="polite">
      <OakFlex $gap="space-between-xs" role="alert">
        <OakRoundIcon
          iconName={feedback === "correct" ? "tick" : "cross"}
          alt=""
          $pa="inner-padding-none"
          $background={feedback === "correct" ? "icon-success" : "icon-error"}
          $colorFilter="text-inverted"
          $width="all-spacing-6"
          $height="all-spacing-6"
        />
        <OakSpan
          $font="heading-6"
          $color={feedback === "correct" ? "text-success" : "text-error"}
          role="alert"
        >
          {feedbackLabel}
        </OakSpan>
      </OakFlex>
      {answerFeedback && (
        <OakSpan
          as="p"
          $mt="space-between-xs"
          $font={feedback === "correct" ? "body-2-bold" : "body-2"}
          role="alert"
        >
          {answerFeedback}
        </OakSpan>
      )}
    </OakBox>
  );
};
