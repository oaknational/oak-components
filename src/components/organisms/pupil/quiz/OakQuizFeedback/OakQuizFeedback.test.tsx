import React from "react";
import "@testing-library/jest-dom";

import { OakQuizFeedback } from "./OakQuizFeedback";

import renderWithTheme from "@/test-helpers/renderWithTheme";

describe(OakQuizFeedback, () => {
  it("matches snapshot", () => {
    const { container } = renderWithTheme(
      <>
        <OakQuizFeedback feedback="correct" answerFeedback="Well done!" />
        <OakQuizFeedback feedback="incorrect" answerFeedback="Keep trying" />
        <OakQuizFeedback feedback="correct" answerFeedback="Nearly there" />
      </>,
    );

    expect(container).toMatchSnapshot();
  });
});
