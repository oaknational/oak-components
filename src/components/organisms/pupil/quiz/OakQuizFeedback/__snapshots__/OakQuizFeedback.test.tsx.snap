// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OakQuizFeedback matches snapshot 1`] = `
[
  .c0 {
  font-family: --var(google-font),Lexend,sans-serif;
}

.c2 {
  width: 1.5rem;
  height: 1.5rem;
  padding: 0rem;
  background: #287c34;
  border-radius: 6.25rem;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c3 {
  position: relative;
  width: 100%;
  min-width: 100%;
  height: 100%;
  min-height: 100%;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c4 {
  -webkit-filter: invert(98%) sepia(98%) saturate(0%) hue-rotate(328deg) brightness(102%) contrast(102%);
  filter: invert(98%) sepia(98%) saturate(0%) hue-rotate(328deg) brightness(102%) contrast(102%);
  object-fit: contain;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.75rem;
}

.c5 {
  color: #287c34;
  font-family: --var(google-font),Lexend,sans-serif;
  font-weight: 600;
  font-size: 1.25rem;
  line-height: 1.5rem;
  -webkit-letter-spacing: 0.0115rem;
  -moz-letter-spacing: 0.0115rem;
  -ms-letter-spacing: 0.0115rem;
  letter-spacing: 0.0115rem;
}

.c6 {
  margin-top: 0.75rem;
  font-family: --var(google-font),Lexend,sans-serif;
  font-weight: 700;
  font-size: 1rem;
  line-height: 1.5rem;
  -webkit-letter-spacing: -0.005rem;
  -moz-letter-spacing: -0.005rem;
  -ms-letter-spacing: -0.005rem;
  letter-spacing: -0.005rem;
}

<div
    aria-live="polite"
    className="c0"
  >
    <div
      className="c0 c1"
      role="alert"
    >
      <div
        className="c2"
      >
        <div
          className="c3"
        >
          <img
            alt=""
            className="c4"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            src="https://res.cloudinary.com/mock-cloudinary-cloud/image/upload/v1699954310/icons/efd3esaor6zqk7seh6kt.svg"
            style={
              {
                "bottom": 0,
                "color": "transparent",
                "height": "100%",
                "left": 0,
                "objectFit": undefined,
                "objectPosition": undefined,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": "100%",
              }
            }
          />
        </div>
      </div>
      <span
        className="c5"
        role="alert"
      >
        Correct
      </span>
    </div>
    <p
      className="c6"
      role="alert"
    >
      Well done!
    </p>
  </div>,
  .c0 {
  font-family: --var(google-font),Lexend,sans-serif;
}

.c3 {
  position: relative;
  width: 100%;
  min-width: 100%;
  height: 100%;
  min-height: 100%;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c2 {
  width: 1.5rem;
  height: 1.5rem;
  padding: 0rem;
  background: #dd0035;
  border-radius: 6.25rem;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c4 {
  -webkit-filter: invert(98%) sepia(98%) saturate(0%) hue-rotate(328deg) brightness(102%) contrast(102%);
  filter: invert(98%) sepia(98%) saturate(0%) hue-rotate(328deg) brightness(102%) contrast(102%);
  object-fit: contain;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.75rem;
}

.c5 {
  color: #dd0035;
  font-family: --var(google-font),Lexend,sans-serif;
  font-weight: 600;
  font-size: 1.25rem;
  line-height: 1.5rem;
  -webkit-letter-spacing: 0.0115rem;
  -moz-letter-spacing: 0.0115rem;
  -ms-letter-spacing: 0.0115rem;
  letter-spacing: 0.0115rem;
}

.c6 {
  margin-top: 0.75rem;
  font-family: --var(google-font),Lexend,sans-serif;
  font-weight: 300;
  font-size: 1rem;
  line-height: 1.5rem;
  -webkit-letter-spacing: -0.005rem;
  -moz-letter-spacing: -0.005rem;
  -ms-letter-spacing: -0.005rem;
  letter-spacing: -0.005rem;
}

<div
    aria-live="polite"
    className="c0"
  >
    <div
      className="c0 c1"
      role="alert"
    >
      <div
        className="c2"
      >
        <div
          className="c3"
        >
          <img
            alt=""
            className="c4"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            src="https://res.cloudinary.com/mock-cloudinary-cloud/image/upload/v1699895179/icons/xigimrbivcaxt4omxamp.svg"
            style={
              {
                "bottom": 0,
                "color": "transparent",
                "height": "100%",
                "left": 0,
                "objectFit": undefined,
                "objectPosition": undefined,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": "100%",
              }
            }
          />
        </div>
      </div>
      <span
        className="c5"
        role="alert"
      >
        Incorrect
      </span>
    </div>
    <p
      className="c6"
      role="alert"
    >
      Keep trying
    </p>
  </div>,
  .c0 {
  font-family: --var(google-font),Lexend,sans-serif;
}

.c2 {
  width: 1.5rem;
  height: 1.5rem;
  padding: 0rem;
  background: #287c34;
  border-radius: 6.25rem;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c3 {
  position: relative;
  width: 100%;
  min-width: 100%;
  height: 100%;
  min-height: 100%;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c4 {
  -webkit-filter: invert(98%) sepia(98%) saturate(0%) hue-rotate(328deg) brightness(102%) contrast(102%);
  filter: invert(98%) sepia(98%) saturate(0%) hue-rotate(328deg) brightness(102%) contrast(102%);
  object-fit: contain;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.75rem;
}

.c5 {
  color: #287c34;
  font-family: --var(google-font),Lexend,sans-serif;
  font-weight: 600;
  font-size: 1.25rem;
  line-height: 1.5rem;
  -webkit-letter-spacing: 0.0115rem;
  -moz-letter-spacing: 0.0115rem;
  -ms-letter-spacing: 0.0115rem;
  letter-spacing: 0.0115rem;
}

.c6 {
  margin-top: 0.75rem;
  font-family: --var(google-font),Lexend,sans-serif;
  font-weight: 700;
  font-size: 1rem;
  line-height: 1.5rem;
  -webkit-letter-spacing: -0.005rem;
  -moz-letter-spacing: -0.005rem;
  -ms-letter-spacing: -0.005rem;
  letter-spacing: -0.005rem;
}

<div
    aria-live="polite"
    className="c0"
  >
    <div
      className="c0 c1"
      role="alert"
    >
      <div
        className="c2"
      >
        <div
          className="c3"
        >
          <img
            alt=""
            className="c4"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            src="https://res.cloudinary.com/mock-cloudinary-cloud/image/upload/v1699954310/icons/efd3esaor6zqk7seh6kt.svg"
            style={
              {
                "bottom": 0,
                "color": "transparent",
                "height": "100%",
                "left": 0,
                "objectFit": undefined,
                "objectPosition": undefined,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": "100%",
              }
            }
          />
        </div>
      </div>
      <span
        className="c5"
        role="alert"
      >
        Correct
      </span>
    </div>
    <p
      className="c6"
      role="alert"
    >
      Nearly there
    </p>
  </div>,
]
`;
